tools/common/Rewriter.cs

```diff
@@ -18,35 +20,54 @@ public class Rewriter {
 		const string classPtrName = "class_ptr";
 		CSToObjCMap map;
 		string pathToXamarinAssembly;
-		string [] assembliesToPatch;
-		string outputDirectory;
-		SimpleAssemblyResolver resolver;
+		string? outputDirectory = null;
 		Dictionary<string, FieldDefinition> csTypeToFieldDef = new Dictionary<string, FieldDefinition> ();
+		IEnumerable<AssemblyDefinition> assemblies;
+		AssemblyDefinition xamarinAssembly;
+		Xamarin.Tuner.DerivedLinkContext linkContext;
 
-		public Rewriter (CSToObjCMap map, string pathToXamarinAssembly, string [] assembliesToPatch, string outputDirectory)
+		public Rewriter (CSToObjCMap map, IEnumerable<AssemblyDefinition> assembliesToPatch, Xamarin.Tuner.DerivedLinkContext? linkContext)
 		{
 			this.map = map;
-			this.pathToXamarinAssembly = pathToXamarinAssembly;
-			this.assembliesToPatch = assembliesToPatch;
-			this.outputDirectory = outputDirectory;
-			resolver = new SimpleAssemblyResolver (assembliesToPatch);
+			this.assemblies = assembliesToPatch;
+			var xasm = assembliesToPatch.Select (assem => assem.MainModule).FirstOrDefault (ContainsNativeHandle)?.Assembly;
+			if (xasm is null) {
+				throw new Exception ("Unable to find Xamarin assembly.");
+			} else {
+				xamarinAssembly = xasm;
+				pathToXamarinAssembly = xamarinAssembly.MainModule.FileName;
+			}
+			if (linkContext is null) {
+				throw new Exception ("Rewriter needs a valid link context.");
+			} else {
+				this.linkContext = linkContext;
+			}
+
 		}
 
-		public void Process ()
+		public string Process ()
 		{
-			var classMap = CreateClassHandles ();
+			Dictionary<string, FieldDefinition> classMap;
+			try {
+				classMap = CreateClassHandles ();
+			} catch (Exception e) {
+				// if this throws, no changes are made to the assemblies
+				// so it's safe to log it on the far side.
+				return e.Message;
+			}
```

Comment:
Any reason it's not a failure to fail to apply the class handle optimization? It should be easy enough for customers to turn it off if need be, and just printing a log message if something goes wrong feels like nobody will notice if it ever breaks.

Reply:
A couple reasons: if this throws an exception, it means that *we've* done something horribly, horribly wrong. It means that someone removed the `ClassHandles` types or its `InitializeClassHandles` method, or `NativeHandle` or any of a number of things that need to be in place for the rest of the code to work. And while this seems like a "Holy cow, we should stop this immediately!" error - it actually isn't. The rest of the process won't run and no changes will have been made to the assemblies and the user is left with a build without `NativeHandle`.